source/client/service_impl.cc

```diff
@@ -0,0 +1,103 @@
+  auto logging_context = std::make_unique<Envoy::Logger::Context>(
+      spdlog::level::from_str(options->verbosity()), "[%T.%f][%t][%L] %v", log_lock_);
+  auto formatter = output_format_factory.create();
+  if (process.run(*formatter)) {
+    response.set_success(true);
+  } else {
+    response.set_error_message("Unkown failure");
+  }
+  *(response.mutable_output()) = formatter->toProto();
+  writeResponseAndFinish(response);
+}
+
+void ServiceImpl::writeResponseAndFinish(const nighthawk::client::ExecutionResponse& response) {
+  response_history_.emplace_back(response);
+  if (!stream_->Write(response)) {
+    ENVOY_LOG(warn, "Stream write failed");
+  }
+  running_ = false;
+}
+
+void ServiceImpl::collectErrorsFromHistory(std::list<std::string>& error_messages) const {
+  for (const auto& result : response_history_) {
+    if (!result.success()) {
+      error_messages.push_back(result.error_message());
+    }
+  }
+}
+
+void ServiceImpl::waitForRunnerThreadCompletion() {
+  if (runner_thread_.joinable()) {
+    runner_thread_.join();
+  }
+  RELEASE_ASSERT(!running_, "running_ ought to be unset");
+}
+
+// TODO(oschaaf): implement a way to cancel test runs, and update rps config on the fly.
+// TODO(oschaaf): unit-test Process, create MockProcess & use in service_test.cc / client_test.cc
+// TODO(oschaaf): should we merge incoming request options with defaults?
+// TODO(oschaaf): aggregate the client's logs and forward them in the grpc response.
+::grpc::Status ServiceImpl::sendCommand(
+    ::grpc::ServerContext* /*context*/,
+    ::grpc::ServerReaderWriter<::nighthawk::client::ExecutionResponse,
+                               ::nighthawk::client::ExecutionRequest>* stream) {
+  std::list<std::string> error_messages;
+  nighthawk::client::ExecutionRequest request;
+  stream_ = stream;
+  while (stream->Read(&request)) {
+    ENVOY_LOG(debug, "Read ExecutionRequest data: {}", request.DebugString());
+    if (request.has_start_request()) {
+      if (running_) {
+        error_messages.emplace_back("Only a single benchmark session is allowed at a time.");
+        break;
+      } else {
+        waitForRunnerThreadCompletion();
+        running_ = true;
+        runner_thread_ = std::thread(&ServiceImpl::handleExecutionRequest, this, request);
+      }
+    } else if (request.has_update_request() || request.has_cancellation_request()) {
+      error_messages.emplace_back("Request is not supported yet.");
+    } else {
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+  waitForRunnerThreadCompletion();
+  collectErrorsFromHistory(error_messages);
+  if (error_messages.size() == 0) {
+    return grpc::Status::OK;
+  }
+
+  std::string error_message;
+  for (const auto& error : error_messages) {
+    error_message = fmt::format("{}\n", error);
+  }
+  ENVOY_LOG(error, "One or more errors processing grpc request stream: {}", error_message);
+  return grpc::Status(grpc::StatusCode::INTERNAL, fmt::format("Error: {}", error_message));
```

Comment:
I think it would be fine to always reply `grpc::Status::OK`, since we have returned errors explicitly as responses on the stream. The way I think of this is that there are CLI execution errors and gRPC transport errors, in this case we have a good gRPC stream, just application level failure. This avoids having to accumulate all the errors, they can be logged as we go.

Reply:
makes a lot of sense, adjusted accordingly.